### PR TITLE
chore: Centralize Status Management

### PR DESCRIPTION
**Pull request by gatici**
# Description

This PR uses new ops feature CollectStatus Event to manage status Charm. Event named collect_unit_status sets the status of charm automatically at the end of every hook.

Reference:

    https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent
    https://discourse.charmhub.io/t/how-to-set-a-charms-status/11771

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

**Review by ghislainbourgeois — APPROVED** [review]
I left only a small suggestion to fix a typo, otherwise LGTM.

File: src/charm.py [hunk 1]
```diff
@@ -72,57 +72,177 @@ def __init__(self, *args) -> None:
...
+
         if not self._nrf_data_is_available:
-            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.add_status(WaitingStatus("Waiting for NRF data to be available"))
             return
+
         if not self._container.exists(path=CONFIG_DIR):
-            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.add_status(WaitingStatus("Waiting for storage to be attached"))
             return
```
Comment (Gmerold) [inline src/charm.py:105]:
```suggestion
            return
            
```

File: src/charm.py [hunk 2]
```diff
@@ -72,57 +72,177 @@ def __init__(self, *args) -> None:
...
 
         Args:
-            event (EventBase): Juju event
+            event: CollectStatusEvent
         """
+        if not self.unit.is_leader():
+            # NOTE: In cases where leader status is lost before the charm is
+            # finished processing all teardown events, this prevents teardown
+            # event code from running. Luckily, for this charm, none of the
+            # teardown code is necessary to preform if we're removing the
```
[the remark this answers is not in the document]
Comment (gatici) [reply src/charm.py]: Fixed the typo.

File: src/charm.py [hunk 3]
```diff
@@ -72,57 +72,177 @@ def __init__(self, *args) -> None:
...
+        if not self._ausf_service_is_running():
+            event.add_status(WaitingStatus("Waiting for AUSF service to start"))
+            return
+
+        event.add_status(ActiveStatus())
+
+    def _ausf_service_is_running(self) -> bool:
+        """Check if the AUSF service is running."""
+        try:
+            self._container.get_service(service_name=self._service_name)
```
Comment (Gmerold) [inline src/charm.py:123]: Does this really check whether the service is running or only is it defined in a pebble plan?
I see that pebble has a method called `is_running()` for checking the service state.

File: src/charm.py [hunk 4]
```diff
@@ -140,7 +260,7 @@ def _on_nrf_broken(self, event: EventBase) -> None:
         Args:
             event (NRFBrokenEvent): Juju event
         """
-        self.unit.status = BlockedStatus("Waiting for fiveg_nrf relation")
+        logger.info("Waiting for fiveg_nrf relation")
```
Comment (Gmerold) [inline src/charm.py:263]: Do we need this method at all? Can we just run the main callback method which will put the charm in the right status?

**Review by Gmerold — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 1 above)
Comment (gatici) [reply src/charm.py:105]: OK

File: src/charm.py (hunk 3 above)
Comment (gatici) [reply src/charm.py:123]: Called is_running() to check the service status.

File: src/charm.py (hunk 1 above)
Comment (gatici) [reply src/charm.py:105]: Done

File: src/charm.py (hunk 4 above)
Comment (gatici) [reply src/charm.py]: Removed this method.

File: src/charm.py (hunk 3 above)
Comment (gatici) [reply src/charm.py]: Done

**Review by Gmerold — APPROVED** [review]
LGTM

**Pull request merged**